bootstrap: translate Windows paths in a way that works for both Cygwin and MSYS2

jeremyd2019 opened the pull request:
Cygwin defaults to rooting Windows paths in /cygdrive/X, while MSYS2 configures them to be /X.  Regardless of configuration, drives are always accessible as /proc/cygdrive/X, so use that.

If there are other shells on Windows that are supported and use /X style paths, perhaps something more complicated needs to be done.

r? @jieyouxu 

@Berrysoft @mati865

jeremyd2019 commented:
https://github.com/rust-lang/rust/issues/137819#issuecomment-2907007888

What confused me is why the `is_cygwin` path wasn't being hit on Cygwin.  Turns out: https://github.com/rust-lang/rust/blob/88b3b520e852e01970c3f519339ba64ed3e7db6d/src/bootstrap/src/core/config/config.rs#L581-L593

It can only be considered `cygwin` if the target has `windows` in it, and it seems the Cygwin target went with `x86_64-pc-cygwin` rather than `x86_64-pc-windows-cygnus` which is what LLVM normalizes a Cygwin target to.

Berrysoft commented:
`is_cygwin` is older than cygwin target. It was used to detect MSYS2 on Windows, to generate and call bash scripts.

Berrysoft commented:
As for target name, Cygwin was decided to be a unix target rather than windows target in the beginning: #134999 . We wanted to reuse most code path for unix, and it seems to be a success now.

mati865 commented:
> If there are other shells on Windows that are supported and use /X style paths, perhaps something more complicated needs to be done.

Haven't heard of any.

jeremyd2019 committed:
bootstrap: translate Windows paths in a way that works for both Cygwin and MSYS2

Cygwin defaults to rooting Windows paths in /cygdrive/X, while MSYS2
configures them to be /X.  Regardless of configuration, drives are
always accessible as /proc/cygdrive/X, so use that.

jieyouxu reviewed (approved):
Thanks

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 108c16eebd1d3de3641c9cdec48314596d01b1b8 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 108c16eebd1d3de3641c9cdec48314596d01b1b8 -->
<!-- homu: {"type":"Approved","sha":"108c16eebd1d3de3641c9cdec48314596d01b1b8","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
